reports/2017-02-24.md

```diff
@@ -0,0 +1,67 @@
+# Development Report for Feb 24, 2017
+
+## containerd Summit
+
+Yesterday, Feb. 23, we had the containerd summit for contributors and maintainers.  We started off by getting everyone up to speed on the project, roadmap, and goals before diving down into specific issues and design of containerd.  We should have the videos up soon for the various sessions and Q&A.
+
+## Runtime Breakout Session
+
+### Shim
+
+We discussed the use of the containerd shim, the costs that it adds, as well as its benefits.  Overall, most of us agreed that the extra memory overhead is worth it for the feature set that it provides.
+
+However, we did make a few changes to the shim in the master branch; a grpc API so that the shim is straightforward to interact with and much more robust in terms of error handling.  We also have one shim per container vs the one shim per process, which was the current implementation.
+
+There were a couple of additional changes we discussed.  One being launching the shim inside the cgroup of the container so that any IO, CPU, and memory that the shim consumes is "charged" to the container.  I'll expand a little more on why this matters in the next section when it comes to logging.
+
+
+### Logging
+
+Logging is a big issue for containers and it’s not exactly straightforward.  Since containers can be used in many different ways the support of logging and connecting to a containers interactively can add complexity to the runtime.  Right now, the shim is responsible for handling the IO of a container and making sure that it is kept open in the event that containerd ( the daemon ) or clients above containerd crash and need to reconnect to containers that they launched.
+
+You can use fifos for reconnecting to container logs but there is concerns about the buffer limit when nothing is reading from the fifo.  If the fifo fills up, then the application in the container will block.  In the end, most container systems log to files on disk but in various formats.  The only part of the container's IO that needs to be a pipe, or fifo for reconnection is STDIN.  STDOUT and STDERR can go directly to a file in most cases.  However, logging the raw output for a container directly to a file is not ideal as both Docker and Kube have specific on disk formats.  
+
+So our initial idea that was discussed is to move some type of "log formatter" into the shim.  This will allow a single container to have its logs written directly to disk from the shim without using fifos for OUT and ERR.  The other problem that this would solve is when a single container has excessive logs.  If a single daemon is writing all the logs to disk, you will see high CPU on the entire daemon.  If we place the shim in the cgroup for the container, then only that container is affected when it starts to log excessively without affecting the performance of the entire system.  The CPU and IO could be charged to the container.
```

Comment:
Have we considered launching a process per container(-shim) that consumes the logs? This is how projects like daemontools and runit avoid overflowing disk space.